File: packages/oxlint-plugin-react-doctor/src/plugin/rules/r3f/r3f-no-sync-readback-in-use-frame.ts
import type { ScopeAnalysis } from "../../semantic/scope-analysis.js";
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNode } from "../../utils/es-tree-node.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { getStaticPropertyName } from "../../utils/get-static-property-name.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import type { RuleContext } from "../../utils/rule-context.js";
import { stripParenExpression } from "../../utils/strip-paren-expression.js";
import { isR3fCallbackStateProperty } from "./utils/is-r3f-callback-state-property.js";
import { resolveR3fCallback } from "./utils/resolve-r3f-callback.js";
import { walkFunctionExecution } from "./utils/walk-function-execution.js";

const CPU_TYPED_ARRAY_CONSTRUCTORS = new Set([
  "BigInt64Array",
  "BigUint64Array",
  "Float32Array",
  "Float64Array",
  "Int8Array",
  "Int16Array",
  "Int32Array",
  "Uint8Array",
  "Uint8ClampedArray",
  "Uint16Array",
  "Uint32Array",
]);
const CANVAS_2D_CONTEXT_NAMES = new Set(["2d"]);
const WEBGL_CONTEXT_NAMES = new Set(["experimental-webgl", "webgl", "webgl2"]);

const isContextFromGetContext = (
  expression: EsTreeNode,
  contextNames: ReadonlySet<string>,
  scopes: ScopeAnalysis,
  visitedSymbolIds: Set<number> = new Set(),
): boolean => {
  const candidate = stripParenExpression(expression);
  if (isNodeOfType(candidate, "Identifier")) {
    const symbol = scopes.symbolFor(candidate);
    if (
      symbol?.kind !== "const" ||
      !symbol.initializer ||
      visitedSymbolIds.has(symbol.id) ||
      symbol.references.some((reference) => reference.flag !== "read")
    ) {
      return false;
    }
    visitedSymbolIds.add(symbol.id);
    return isContextFromGetContext(symbol.initializer, contextNames, scopes, visitedSymbolIds);
  }
  if (
    !isNodeOfType(candidate, "CallExpression") ||
    !isNodeOfType(candidate.callee, "MemberExpression") ||
    getStaticPropertyName(candidate.callee) !== "getContext"
  ) {
    return false;
  }
  const contextName = candidate.arguments[0];
  if (!contextName || isNodeOfType(contextName, "SpreadElement")) return false;
  const staticContextName = stripParenExpression(contextName);
  return Boolean(
    isNodeOfType(staticContextName, "Literal") &&
    typeof staticContextName.value === "string" &&
    contextNames.has(staticContextName.value),
  );
};

const isCpuTypedArray = (
  expression: EsTreeNode,
  scopes: ScopeAnalysis,
  visitedSymbolIds: Set<number> = new Set(),
): boolean => {
  const candidate = stripParenExpression(expression);
  if (isNodeOfType(candidate, "Identifier")) {
    const symbol = scopes.symbolFor(candidate);
    if (
      symbol?.kind !== "const" ||
      !symbol.initializer ||
      visitedSymbolIds.has(symbol.id) ||
      symbol.references.some((reference) => reference.flag !== "read")
    ) {
      return false;
    }
    visitedSymbolIds.add(symbol.id);
    return isCpuTypedArray(symbol.initializer, scopes, visitedSymbolIds);
  }
  if (!isNodeOfType(candidate, "NewExpression")) return false;
  const callee = stripParenExpression(candidate.callee);
  return (
    isNodeOfType(callee, "Identifier") &&
    CPU_TYPED_ARRAY_CONSTRUCTORS.has(callee.name) &&
    scopes.isGlobalReference(callee)
  );
};

const getReadbackKind = (
  node: EsTreeNodeOfType<"CallExpression">,
  callback: EsTreeNode,
  context: RuleContext,
): "canvas" | "three" | "webgl" | null => {
  if (!isNodeOfType(node.callee, "MemberExpression")) return null;
  const methodName = getStaticPropertyName(node.callee);
  if (
    methodName === "readRenderTargetPixels" &&
    (isR3fCallbackStateProperty(node.callee.object, callback, "gl", context.scopes) ||
      isR3fCallbackStateProperty(node.callee.object, callback, "renderer", context.scopes))
  ) {
    return "three";
  }
  if (
    methodName === "getImageData" &&
    isContextFromGetContext(node.callee.object, CANVAS_2D_CONTEXT_NAMES, context.scopes)
  ) {
    return "canvas";
  }
  if (
    methodName === "readPixels" &&
    isContextFromGetContext(node.callee.object, WEBGL_CONTEXT_NAMES, context.scopes)
  ) {
    const destination = node.arguments[6];
    return destination &&
      !isNodeOfType(destination, "SpreadElement") &&
      isCpuTypedArray(destination, context.scopes)
      ? "webgl"
      : null;
  }
  return null;
};

export const r3fNoSyncReadbackInUseFrame = defineRule({
  id: "r3f-no-sync-readback-in-use-frame",
  title: "Synchronous readback inside useFrame",
  severity: "warn",
  recommendation:
    "Move readback to a discrete or serialized async path and reuse the latest completed result during frames",
  create: (context: RuleContext) => ({
    CallExpression(node: EsTreeNodeOfType<"CallExpression">) {
      const callback = resolveR3fCallback(node, "useFrame", context.scopes);
      if (!callback) return;
      walkFunctionExecution(callback, context.scopes, (candidate, isConditionallyExecuted) => {
        if (!isNodeOfType(candidate, "CallExpression")) return;
        const readbackKind = getReadbackKind(candidate, callback, context);
        if (!readbackKind) return;
        if (isConditionallyExecuted) return;
        context.report({
          node: candidate,
          message:
            readbackKind === "canvas"
              ? "getImageData copies pixels to the CPU on every frame. Sample on demand or at a lower rate"
              : "Synchronous GPU readback can stall the frame until prior GPU work completes. Use an asynchronous or event-driven readback path",
        });
      });
    },
  }),
});

File: native/oxlint/rules/r3f-no-sync-readback-in-use-frame.rs
use oxc_ast::AstKind;
use oxc_diagnostics::OxcDiagnostic;
use oxc_span::GetSpan;

use crate::{
    context::LintContext,
    rule::{Rule, RuleCategory, RuleInfo, RuleMeta},
};

const CANVAS_MESSAGE: &str =
    "getImageData copies pixels to the CPU on every frame. Sample on demand or at a lower rate";
const GPU_MESSAGE: &str = "Synchronous GPU readback can stall the frame until prior GPU work completes. Use an asynchronous or event-driven readback path";
const R3F_SYNC_READBACK_PUBLIC_MODULES: [&str; 5] = [
    "@react-three/fiber",
    "@react-three/fiber/legacy",
    "@react-three/fiber/native",
    "@react-three/fiber/webgpu",
    "react-three-fiber",
];

#[derive(Debug, Default, Clone)]
pub struct R3FNoSyncReadbackInUseFrame;

impl RuleMeta for R3FNoSyncReadbackInUseFrame {
    const NAME: &'static str = "r3f-no-sync-readback-in-use-frame";
    const PLUGIN: &'static str = "react_doctor_native";
    const CATEGORY: RuleCategory = RuleCategory::Perf;
    const VERSION: &'static str = "0.1.0";
    const INFO: RuleInfo = RuleInfo {
        short_description: "Disallow synchronous readback inside useFrame.",
    };
}

impl Rule for R3FNoSyncReadbackInUseFrame {
    fn run_once<'a>(&self, ctx: &LintContext<'a>) {
        let analysis = build_possible_static_property_write_analysis(ctx);
        let node_index = build_local_callback_nearest_function_node_index(ctx);
        let mut resolution_cache = LocalFunctionResolutionCache::default();
        for node in ctx.nodes().iter() {
            let AstKind::CallExpression(call_expression) = node.kind() else {
                continue;
            };
            if !module_api_reference_matches(
                &call_expression.callee,
                "useFrame",
                &R3F_SYNC_READBACK_PUBLIC_MODULES,
                &analysis,
                ctx,
            ) && !type_import_module_api_reference_matches(
                &call_expression.callee,
                "useFrame",
                &R3F_SYNC_READBACK_PUBLIC_MODULES,
                &analysis,
                ctx,
            ) {
                continue;
            }
            let Some(callback_expression) = call_expression
                .arguments
                .first()
                .and_then(oxc_ast::ast::Argument::as_expression)
            else {
                continue;
            };
            let Some(callback_id) = resolve_r3f_analyzed_callback_function_id(
                callback_expression,
                &analysis,
                ctx,
                &mut resolution_cache,
            ) else {
                continue;
            };
            if matches!(
                ctx.nodes().get_node(callback_id).kind(),
                AstKind::Function(function) if function.generator
            ) {
                continue;
            }
            for_each_analyzed_synchronous_execution_node(
                callback_id,
                &analysis,
                &node_index,
                ctx,
                &mut resolution_cache,
                |candidate, callback_id, is_conditionally_executed, _| {
                    if is_conditionally_executed {
                        return;
                    }
                    let AstKind::CallExpression(readback_call) = candidate.kind() else {
                        return;
                    };
                    let Some(member_expression) = readback_call.callee.as_member_expression()
                    else {
                        return;
                    };
                    let message = match static_member_expression_property_name(member_expression) {
                        Some("readRenderTargetPixels")
                            if r3f_callback_state_property_matches(
                                member_expression.object(),
                                callback_id,
                                "gl",
                                ctx,
                            ) || r3f_callback_state_property_matches(
                                member_expression.object(),
                                callback_id,
                                "renderer",
                                ctx,
                            ) =>
                        {
                            GPU_MESSAGE
                        }
                        Some("getImageData")
                            if is_context_from_get_context(
                                member_expression.object(),
                                &["2d"],
                                ctx,
                            ) =>
                        {
                            CANVAS_MESSAGE
                        }
                        Some("readPixels")
                            if is_webgl_context_reference(member_expression.object(), ctx)
                                && readback_call
                                    .arguments
                                    .get(6)
                                    .and_then(oxc_ast::ast::Argument::as_expression)
                                    .is_some_and(|destination| {
                                        is_cpu_typed_array(destination, ctx)
                                    }) =>
                        {
                            GPU_MESSAGE
                        }
                        _ => return,
                    };
                    ctx.diagnostic(OxcDiagnostic::warn(message).with_label(candidate.span()));
                },
            );
        }
    }
}
